src/object/obj_tx.c

```diff
@@ -1899,8 +1915,8 @@ dc_tx_restart(tse_task_t *task)
 		D_GOTO(out_task, rc = -DER_NO_HDL);
 
 	D_MUTEX_LOCK(&tx->tx_lock);
-	if (tx->tx_status != TX_OPEN && tx->tx_status != TX_FAILED) {
-		D_ERROR("Can't restart non-open/non-failed state TX (%d)\n",
+	if (tx->tx_status != TX_FAILED) {
```

Comment:
If the application want to restart the TX based on some conditional operation result, how to process? have to re-open the transaction? That seems inconvenient.

Reply:
If the application hasn't encountered any -DER_TX_RESTART, and it wants to restart the TX, it likely wants also to read or overwrite any other updates committed after the original open and before the restart. Of course, this is essentially my guess.

If the guess has some truth, I think we should require the application to use a different restart than the restart for -DER_TX_RESTARTs. This can be done by

- close and open,
- a new daos_tx_reopen,
- daos_tx_restart plus an internal check to see if the TX has encountered any -DER_TX_RESTART, or
- daos_tx_restart plus a flag.

This PR assumes the first two options, but I'm open to the other options too.

Reply:
@mchaarawi, @gnailzenh, @johannlombardi, which option would you prefer? See my last comment.

Reply:
The 3rd option above is related to storing the error in dc_tx so that the error can be retrieved by the TX handle. We briefly discussed that before, but there was no conclusion yet.

Reply:
Sorry, I do not understand why need to restrict to restart an TX in open status? What is the benefit for that?

Reply:
It is neither related with optimization nor correctness, just some semantics definition or restriction, then I also would like to see others option for that.

Reply:
Option 1 is OK from my perspective but not too convenient.
It would be more convenient from a programming perspective to support option 3?
maybe a question to you and Fan Yong: Is there any challenge or gotchas to support option 3 really?

Reply:
My understand is that it is only for semantics restriction to avoid potential trouble. As Liwei said: we can explain to the user that tx_restart() does not guarantee the current TX can see others new committed results, but the user may still call tx_restart() with such expectation. So he prefers to forbid tx_restart() if it is not required by the server. But I think that since we have already claimed related things, the user should not expect the things that DAOS does not promise.

Reply:
@Nasf-Fan, after a discussion on Wednesday, we decided to restrict daos_tx_restart to -DER_TX_RESTART cases for now; if an alternative to "daos_tx_close-daos_tx_open" proves necessary, we can revisit the API.

Reply:
In current implementation, there are still some other cases to set "tx_state" as "TX_FAILED" even if the server does not return "-DER_TX_RESTART", please check dc_tx_commit_cb(). So here, we do not prevent the application restarting the TX that is not required the server. Then the implementation does not match the expected semantics.

Reply:
@Nasf-Fan, I've converted those to TX_ABORTED for errors other than -DER_TX_RESTART.

Reply:
I am not against to set TX state as "ABORTED" as your patch does, but I am wondering whether we need tx_abort() API or not, since daos client will automatically set the TX as "aborted" when necessary. If the application wants to abort the TX that does not hit any failure, it can directly call tx_close(). I cannot find any special reason for tx_abort() API.

Reply:
I have no strong opinion at the moment. @mchaarawi, @gnailzenh, @johannlombardi?

Reply:
if i remember my last discussion on this, we want to keep aborted API.
I reviewed the patch again, and i'm ok with it as is.
I just want to make sure that TX sets aborted only in the case where tx_commit fails for other than RESTART or retryable error, right?
In that case im OK with this patch

Reply:
> if i remember my last discussion on this, we want to keep aborted API.

Fine with me. @Nasf-Fan, are you okay with keeping daos_tx_abort for now?

> I reviewed the patch again, and i'm ok with it as is.

Thanks a lot.

> I just want to make sure that TX sets aborted only in the case where tx_commit fails for other than RESTART or retryable error, right?

Right, only for the daos_tx_commit code path. An alternative in this case is to keep the TX in COMMITTING state, rather than ABORTED, which would be slightly more consistent with non-daos_tx_commit code paths.